/*
 * This file is part of the https://github.com/QQxiaoming/quardCRT.git
 * project.
 *
 * Copyright (C) 2023 Quard <2014500726@smail.xtu.edu.cn>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */
#include <QSplitter>
#include <QTreeView>
#include <QLabel>
#include <QStringListModel>
#include <QSerialPortInfo>
#include <QFileInfo>
#include <QStandardItemModel>

#include "sessionoptionswindow.h"

#include "ui_sessionoptionswindow.h"
#include "ui_sessionoptionsgeneralwidget.h"

#include "ui_sessionoptionstelnetproperties.h"
#include "ui_sessionoptionsserialproperties.h"
#include "ui_sessionoptionslocalshellproperties.h"
#include "ui_sessionoptionsnamepipeproperties.h"
#include "ui_sessionoptionsrawproperties.h"
#include "ui_sessionoptionsssh2properties.h"
#include "ui_sessionoptionsvncproperties.h"

#include "ui_sessionoptionsserialstate.h"
#include "ui_sessionoptionslocalshellstate.h"

SessionOptionsWindow::SessionOptionsWindow(QWidget *parent) :
    QDialog(parent),
    ui(new Ui::SessionOptionsWindow)
{
    ui->setupUi(this);

    setWindowModality(Qt::ApplicationModal);
    setWindowFlags(Qt::Tool);

    ui->splitter->setHandleWidth(1);
    ui->treeView->setMinimumWidth(140);
    ui->treeView->setHeaderHidden(true);
    model = new QStringListModel(ui->treeView);
    ui->treeView->setModel(model);
    emptyWidget = new QWidget(this);
    ui->stackedWidget->addWidget(emptyWidget);
    ui->splitter->setSizes(QList<int>() << 140 << 500);
    ui->splitter->setCollapsible(0,false);
    ui->splitter->setCollapsible(1,false);

    sessionOptionsGeneralWidget = new SessionOptionsGeneralWidget(this);
    ui->stackedWidget->addWidget(sessionOptionsGeneralWidget);
    sessionOptionsTelnetProperties = new SessionOptionsTelnetProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsTelnetProperties);
    sessionOptionsSerialProperties = new SessionOptionsSerialProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsSerialProperties);
    sessionOptionsLocalShellProperties = new SessionOptionsLocalShellProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsLocalShellProperties);
    sessionOptionsNamePipeProperties = new SessionOptionsNamePipeProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsNamePipeProperties);
    sessionOptionsRawProperties = new SessionOptionsRawProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsRawProperties);
    sessionOptionsSSH2Properties = new SessionOptionsSsh2Properties(this);
    ui->stackedWidget->addWidget(sessionOptionsSSH2Properties);
    sessionOptionsVNCProperties = new SessionOptionsVNCProperties(this);
    ui->stackedWidget->addWidget(sessionOptionsVNCProperties);

    sessionOptionsLocalShellState = new SessionOptionsLocalShellState(this);
    ui->stackedWidget->addWidget(sessionOptionsLocalShellState);
    sessionOptionsSerialState = new SessionOptionsSerialState(this);
    ui->stackedWidget->addWidget(sessionOptionsSerialState);

    setactiveProperties(-1);
    setactiveState(-1);
    ui->stackedWidget->setCurrentWidget(sessionOptionsGeneralWidget);

    retranslateUi();

    connect(ui->treeView, &QTreeView::clicked, [&](const QModelIndex &index) {
        if (index.row() == 0) {
            setactiveProperties(-1);
            setactiveState(-1);
            ui->stackedWidget->setCurrentWidget(sessionOptionsGeneralWidget);
        } else if (index.row() == 1) {
            setactiveState(-1);
            setactiveProperties(sessionOptionsGeneralWidget->ui->comboBoxProtocol->currentIndex());
        } else if (index.row() == 2) {
            setactiveProperties(-1);
            setactiveState(sessionOptionsGeneralWidget->ui->comboBoxProtocol->currentIndex());
        }
    });

    connect(ui->buttonBox, &QDialogButtonBox::accepted, this, &SessionOptionsWindow::buttonBoxAccepted);
    connect(ui->buttonBox, &QDialogButtonBox::rejected, this, &SessionOptionsWindow::buttonBoxRejected);
    
    connect(sessionOptionsSerialProperties->ui->pushButtonRefresh, &QPushButton::clicked, [&]() {
        QSerialPortInfo serialPortInfo;
        sessionOptionsSerialProperties->ui->comboBoxPortName->clear();
        foreach(serialPortInfo, QSerialPortInfo::availablePorts()) {
            if(serialPortInfo.description().isEmpty()) {
                sessionOptionsSerialProperties->ui->comboBoxPortName->addItem(serialPortInfo.portName());
            } else {
                sessionOptionsSerialProperties->ui->comboBoxPortName->addItem(serialPortInfo.portName()+" - "+serialPortInfo.description());
            }
        }
    });
}

SessionOptionsWindow::~SessionOptionsWindow()
{
    delete ui;
}

void SessionOptionsWindow::retranslateUi()
{
    model->setStringList(QStringList() << tr("General") << tr("Properties") << tr("State"));
    ui->retranslateUi(this);
    sessionOptionsGeneralWidget->ui->retranslateUi(this);
    sessionOptionsTelnetProperties->ui->retranslateUi(this);
    sessionOptionsSerialProperties->ui->retranslateUi(this);
    sessionOptionsLocalShellProperties->ui->retranslateUi(this);
    sessionOptionsNamePipeProperties->ui->retranslateUi(this);
    sessionOptionsRawProperties->ui->retranslateUi(this);
    sessionOptionsSSH2Properties->ui->retranslateUi(this);
    sessionOptionsVNCProperties->ui->retranslateUi(this);
    sessionOptionsLocalShellState->ui->retranslateUi(this);
    sessionOptionsLocalShellState->ui->treeViewInfo->model()->setHeaderData(1, Qt::Horizontal, tr("Name"));
    sessionOptionsSerialState->ui->retranslateUi(this);
}

void SessionOptionsWindow::setactiveProperties(int index)
{
    ui->stackedWidget->setCurrentWidget(emptyWidget);
    if(index == -1) {
        return;
    }
    switch(index) {
    case 0:
        ui->stackedWidget->setCurrentWidget(sessionOptionsTelnetProperties);
        break;
    case 1:
        ui->stackedWidget->setCurrentWidget(sessionOptionsSerialProperties);
        break;
    case 2:
        ui->stackedWidget->setCurrentWidget(sessionOptionsLocalShellProperties);
        break;
    case 3:
        ui->stackedWidget->setCurrentWidget(sessionOptionsRawProperties);
        break;
    case 4:
        ui->stackedWidget->setCurrentWidget(sessionOptionsNamePipeProperties);
        break;
    case 5:
        ui->stackedWidget->setCurrentWidget(sessionOptionsSSH2Properties);
        break;
    case 6:
        ui->stackedWidget->setCurrentWidget(sessionOptionsVNCProperties);
        break;
    }
}

void SessionOptionsWindow::setactiveState(int index)
{
    ui->stackedWidget->setCurrentWidget(emptyWidget);

    if(index == -1) {
        return;
    }
    switch(index) {
    case 1:
        ui->stackedWidget->setCurrentWidget(sessionOptionsSerialState);
        break;
    case 2:
        ui->stackedWidget->setCurrentWidget(sessionOptionsLocalShellState);
        break;
    }
}

void SessionOptionsWindow::setSessionProperties(QString name, QuickConnectWindow::QuickConnectData data)
{
    currentSessionName = name;
    sessionOptionsLocalShellState->ui->treeViewInfo->model()->removeRows(0, sessionOptionsLocalShellState->ui->treeViewInfo->model()->rowCount());
    sessionOptionsGeneralWidget->ui->comboBoxProtocol->setCurrentIndex(data.type);
    sessionOptionsGeneralWidget->ui->lineEditName->setText(name);
    switch(data.type) {
    case QuickConnectWindow::Telnet:
        sessionOptionsTelnetProperties->ui->lineEditHostname->setText(data.TelnetData.hostname);
        sessionOptionsTelnetProperties->ui->spinBoxPort->setValue(data.TelnetData.port);
        sessionOptionsTelnetProperties->ui->comboBoxWebSocket->setCurrentIndex(data.TelnetData.webSocket);
        break;
    case QuickConnectWindow::Serial: {
        sessionOptionsSerialProperties->ui->comboBoxPortName->clear();
        QSerialPortInfo serialPortInfo;
        bool match = false;
        foreach(serialPortInfo, QSerialPortInfo::availablePorts()) {
            if(serialPortInfo.description().isEmpty()) {
                sessionOptionsSerialProperties->ui->comboBoxPortName->addItem(serialPortInfo.portName());
            } else {
                sessionOptionsSerialProperties->ui->comboBoxPortName->addItem(serialPortInfo.portName()+" - "+serialPortInfo.description());
            }
            if(serialPortInfo.portName() == data.SerialData.portName) {
                sessionOptionsSerialProperties->ui->comboBoxPortName->setCurrentIndex(sessionOptionsSerialProperties->ui->comboBoxPortName->count()-1);
                match = true;
            }
        }
        if(!match) {
            sessionOptionsSerialProperties->ui->comboBoxPortName->addItem(data.SerialData.portName);
            sessionOptionsSerialProperties->ui->comboBoxPortName->setCurrentText(data.SerialData.portName);
        }
        sessionOptionsSerialProperties->ui->spinBoxBaudRate->setValue(data.SerialData.baudRate);
        sessionOptionsSerialProperties->ui->comboBoxDataBits->setCurrentText(QString::number(data.SerialData.dataBits));
        sessionOptionsSerialProperties->ui->comboBoxParity->setCurrentText(QString::number(data.SerialData.parity));
        sessionOptionsSerialProperties->ui->comboBoxStopBits->setCurrentText(QString::number(data.SerialData.stopBits));
        sessionOptionsSerialProperties->ui->checkBoxFlowControl->setChecked(data.SerialData.flowControl);
        sessionOptionsSerialProperties->ui->checkBoxXEnable->setChecked(data.SerialData.xEnable);
        break;
    }
    case QuickConnectWindow::LocalShell:
        sessionOptionsLocalShellProperties->ui->lineEditCommand->setText(data.LocalShellData.command);
        break;
    case QuickConnectWindow::Raw:
        sessionOptionsRawProperties->ui->lineEditHostname->setText(data.RawData.hostname);
        sessionOptionsRawProperties->ui->spinBoxPort->setValue(data.RawData.port);
        break;
    case QuickConnectWindow::NamePipe:
        sessionOptionsNamePipeProperties->ui->lineEditPipeName->setText(data.NamePipeData.pipeName);
        break;
    case QuickConnectWindow::SSH2:
        sessionOptionsSSH2Properties->ui->lineEditHostname->setText(data.SSH2Data.hostname);
        sessionOptionsSSH2Properties->ui->spinBoxPort->setValue(data.SSH2Data.port);
        sessionOptionsSSH2Properties->ui->lineEditUserName->setText(data.SSH2Data.username);
        sessionOptionsSSH2Properties->lineEditPassword->setText(data.SSH2Data.password);
        sessionOptionsSSH2Properties->lineEditPassword->setPasswordShown(false);
        break;
    case QuickConnectWindow::VNC:
        sessionOptionsVNCProperties->ui->lineEditHostname->setText(data.VNCData.hostname);
        sessionOptionsVNCProperties->ui->spinBoxPort->setValue(data.VNCData.port);
        sessionOptionsVNCProperties->lineEditPassword->setText(data.VNCData.password);
        sessionOptionsVNCProperties->lineEditPassword->setPasswordShown(false);
        break;
    }
}

void SessionOptionsWindow::setSessionState(SessionsWindow::StateInfo state)
{
    auto setLabelState = [](QLabel *label, SessionsWindow::SessionsState st) {
        switch(st) {
            case SessionsWindow::Connected:
                label->setPixmap(QPixmap(QFontIcon::icon(QChar(0xf0c1), Qt::green).pixmap(16,16)));
                break;
            case SessionsWindow::Disconnected:
            case SessionsWindow::Error:
                label->setPixmap(QPixmap(QFontIcon::icon(QChar(0xf127), Qt::red).pixmap(16,16)));
                break;
            case SessionsWindow::Locked:
                label->setPixmap(QPixmap(QFontIcon::icon(QChar(0xf084), Qt::yellow).pixmap(16,16)));
                break;
            case SessionsWindow::BroadCasted:
                label->setPixmap(QPixmap(QFontIcon::icon(QChar(0xf08e), Qt::yellow).pixmap(16,16)));
                break;
        }
    };
    switch(state.type) {
        case SessionsWindow::Serial: {
            auto getSize = [](int64_t size) -> QString {
                if( size <= 1024) {
                    return QString("%1 B").arg(size);
                } else if ( size <= 1024 * 1024 ) {
                    return QString::number(size / 1024.0, 'f', 2) + QString(" KB");
                } else if ( size <= 1024 * 1024 * 1024 ) {
                    return QString::number(size / (1024.0 * 1024.0), 'f', 2) + QString(" MB");
                } else {
                    return QString::number(size / (1024.0 * 1024.0 * 1024.0), 'f', 2) + QString(" GB");
                }
            };
            sessionOptionsSerialState->ui->lineEditRx1->setText(getSize(state.serial.rx_total));
            sessionOptionsSerialState->ui->lineEditTx1->setText(getSize(state.serial.tx_total));
            sessionOptionsSerialState->ui->lineEditRx2->setText(QString::number(state.serial.rx_total)+" B");
            sessionOptionsSerialState->ui->lineEditTx2->setText(QString::number(state.serial.tx_total)+" B");
            setLabelState(sessionOptionsSerialState->ui->labelState, state.state);
            break;
        }
        case SessionsWindow::LocalShell: {
            if(state.state == SessionsWindow::Connected) {
                QStandardItemModel *model = (QStandardItemModel *)sessionOptionsLocalShellState->ui->treeViewInfo->model();
                std::function<void(IPtyProcess::pidTree_t, QStandardItem *)> addNode = [&](IPtyProcess::pidTree_t tree, QStandardItem *parent) {
                    qint64 pid = tree.pidInfo.pid;
                    QString path = tree.pidInfo.command;
                    QFileInfo fileInfo(path);
                    QString name = fileInfo.fileName();
                    QList<QStandardItem *> items = { new QStandardItem(), new QStandardItem() };
                    items[0]->setData(pid, Qt::DisplayRole);
                    items[1]->setData(name, Qt::DisplayRole);
                    items[1]->setData(path, Qt::ToolTipRole);
                    parent->appendRow(items);
                    foreach(IPtyProcess::pidTree_t child, tree.children) {
                        addNode(child, items[0]);
                    }
                };
                addNode(state.localShell.tree, model->invisibleRootItem());
                sessionOptionsLocalShellState->ui->treeViewInfo->expandToDepth(2);
            }
            setLabelState(sessionOptionsLocalShellState->ui->labelState, state.state);
            break;
        }
        default:
            break;
    }
}

void SessionOptionsWindow::setReadOnly(bool enable) {
    auto comboxSetReadOnly = [&](QComboBox *combox, QLineEdit *lineEdit) {
        if(enable) {
            combox->setVisible(false);
            lineEdit->setVisible(true);
            lineEdit->setText(combox->currentText());
            lineEdit->setReadOnly(true);
        } else {
            combox->setVisible(true);
            lineEdit->setVisible(false);
        }
    };
    auto checkboxSetReadOnly = [&](QCheckBox* checkBox) {
        checkBox->setAttribute(Qt::WA_TransparentForMouseEvents, enable);
        checkBox->setFocusPolicy(enable ? Qt::NoFocus : Qt::StrongFocus);
    };
    comboxSetReadOnly(sessionOptionsGeneralWidget->ui->comboBoxProtocol, sessionOptionsGeneralWidget->ui->lineEditReadOnlyProtocol);
    sessionOptionsGeneralWidget->ui->lineEditName->setReadOnly(enable);

    sessionOptionsTelnetProperties->ui->lineEditHostname->setReadOnly(enable);
    sessionOptionsTelnetProperties->ui->spinBoxPort->setReadOnly(enable);
    comboxSetReadOnly(sessionOptionsTelnetProperties->ui->comboBoxWebSocket, sessionOptionsTelnetProperties->ui->lineEditReadOnlyWebSocket);

    sessionOptionsSerialProperties->ui->spinBoxBaudRate->setReadOnly(enable);
    comboxSetReadOnly(sessionOptionsSerialProperties->ui->comboBoxPortName, sessionOptionsSerialProperties->ui->lineEditReadOnlyPortName);
    comboxSetReadOnly(sessionOptionsSerialProperties->ui->comboBoxDataBits, sessionOptionsSerialProperties->ui->lineEditReadOnlyDataBits);
    comboxSetReadOnly(sessionOptionsSerialProperties->ui->comboBoxParity, sessionOptionsSerialProperties->ui->lineEditReadOnlyParity);
    comboxSetReadOnly(sessionOptionsSerialProperties->ui->comboBoxStopBits, sessionOptionsSerialProperties->ui->lineEditReadOnlyStopBits);
    checkboxSetReadOnly(sessionOptionsSerialProperties->ui->checkBoxXEnable);
    checkboxSetReadOnly(sessionOptionsSerialProperties->ui->checkBoxFlowControl);
    sessionOptionsSerialProperties->ui->pushButtonRefresh->setVisible(!enable);

    sessionOptionsLocalShellProperties->ui->lineEditCommand->setReadOnly(enable);

    sessionOptionsRawProperties->ui->lineEditHostname->setReadOnly(enable);
    sessionOptionsRawProperties->ui->spinBoxPort->setReadOnly(enable);

    sessionOptionsNamePipeProperties->ui->lineEditPipeName->setReadOnly(enable);

    sessionOptionsSSH2Properties->ui->lineEditHostname->setReadOnly(enable);
    sessionOptionsSSH2Properties->ui->spinBoxPort->setReadOnly(enable);
    sessionOptionsSSH2Properties->ui->lineEditUserName->setReadOnly(enable);
    sessionOptionsSSH2Properties->lineEditPassword->setReadOnly(enable);

    sessionOptionsVNCProperties->ui->lineEditHostname->setReadOnly(enable);
    sessionOptionsVNCProperties->ui->spinBoxPort->setReadOnly(enable);
    sessionOptionsVNCProperties->lineEditPassword->setReadOnly(enable);
}

void SessionOptionsWindow::buttonBoxAccepted(void)
{
    if(currentSessionName.isEmpty()) {
        return;
    }
    QuickConnectWindow::QuickConnectData data;
    data.type = (QuickConnectWindow::QuickConnectType)sessionOptionsGeneralWidget->ui->comboBoxProtocol->currentIndex();
    switch(data.type) {
    case QuickConnectWindow::Telnet:
        data.TelnetData.hostname = sessionOptionsTelnetProperties->ui->lineEditHostname->text();
        data.TelnetData.port = sessionOptionsTelnetProperties->ui->spinBoxPort->value();
        data.TelnetData.webSocket = sessionOptionsTelnetProperties->ui->comboBoxWebSocket->currentIndex();
        break;
    case QuickConnectWindow::Serial:
        data.SerialData.portName = sessionOptionsSerialProperties->ui->comboBoxPortName->currentText().split(" - ")[0];
        data.SerialData.baudRate = sessionOptionsSerialProperties->ui->spinBoxBaudRate->value();
        data.SerialData.dataBits = sessionOptionsSerialProperties->ui->comboBoxDataBits->currentText().toInt();
        data.SerialData.parity = sessionOptionsSerialProperties->ui->comboBoxParity->currentText().toInt();
        data.SerialData.stopBits = sessionOptionsSerialProperties->ui->comboBoxStopBits->currentText().toInt();
        data.SerialData.flowControl = sessionOptionsSerialProperties->ui->checkBoxFlowControl->isChecked();
        data.SerialData.xEnable = sessionOptionsSerialProperties->ui->checkBoxXEnable->isChecked();
        break;
    case QuickConnectWindow::LocalShell:
        data.LocalShellData.command = sessionOptionsLocalShellProperties->ui->lineEditCommand->text();
        break;
    case QuickConnectWindow::Raw:
        data.RawData.hostname = sessionOptionsRawProperties->ui->lineEditHostname->text();
        data.RawData.port = sessionOptionsRawProperties->ui->spinBoxPort->value();
        break;
    case QuickConnectWindow::NamePipe:
        data.NamePipeData.pipeName = sessionOptionsNamePipeProperties->ui->lineEditPipeName->text();
        break;
    case QuickConnectWindow::SSH2:
        data.SSH2Data.hostname = sessionOptionsSSH2Properties->ui->lineEditHostname->text();
        data.SSH2Data.port = sessionOptionsSSH2Properties->ui->spinBoxPort->value();
        data.SSH2Data.username = sessionOptionsSSH2Properties->ui->lineEditUserName->text();
        data.SSH2Data.password = sessionOptionsSSH2Properties->lineEditPassword->text();
        break;
    case QuickConnectWindow::VNC:
        data.VNCData.hostname = sessionOptionsVNCProperties->ui->lineEditHostname->text();
        data.VNCData.port = sessionOptionsVNCProperties->ui->spinBoxPort->value();
        data.VNCData.password = sessionOptionsVNCProperties->lineEditPassword->text();
        break;
    }
    emit sessionPropertiesChanged(currentSessionName, data, sessionOptionsGeneralWidget->ui->lineEditName->text());
    emit this->accepted();
}

void SessionOptionsWindow::buttonBoxRejected(void)
{
    emit this->rejected();
}

void SessionOptionsWindow::showEvent(QShowEvent *event)
{
    setactiveProperties(-1);
    setactiveState(-1);
    ui->stackedWidget->setCurrentWidget(sessionOptionsGeneralWidget);
    retranslateUi();
    QDialog::showEvent(event);
}
